### PR TITLE
Form translations

### PR DESCRIPTION
**Pull request by ciara-mc**
This PR:
- uses an app version of xlsform template extending from the package version
- contains seeders for languages and language string types
- creates template languages, survey rows and language strings when a new xlsform template is created
- shows the user all languages available for the template
- creates an export with one row per survey row - language string type with all existing translations and a space to add a new translation
- creates a new template language and language strings when a completed translation file is added

To Fix
- [x] pass the translation file to the XlsformTemplateLanguageImport

Further Development
- [x] add validation for translation import - ensure no missing transaltions
- [x] case when template xlsfile is replaced
- [x] enable editing of language strings
- [x] enable picking of language / creation of XlsformTemplateLanguage before uploading translations file

For future PR:
- add instructions in export file/on the relation manager & improve design of export
- multi worksheets - include choices

### CODE REVIEW

**Review by dave-mills — COMMENTED** [review]
On the key issue of getting the uploaded file to the importer: 

- don't use $livewire->data['translation_file']. The form's data is passed to the action function as $data, so it's `$data['translation_file']` - which will give you the filename. 

The other thing that I just learned is that Filament has its own filesystem config item (`config/filament` - which uses the .env variable `FILAMENT_FILESYSTEM_DISK` and defaults to "public" ), but the main Laravel `FILESYSTEM_DISK` defaults to "local". So, without changing either of those, the FileUpload will put the file into the public disk, and then the `Storage::path()` will look in the local disk. 

TLDR - set `FILESYSTEM_DISK` and `FILAMENT_FILESYSTEM_DISK` to the same value in .env, and then you can do: 

```
$uploadedFile = $data['translation_file'];

$file = Storage::path($uploadedFile);

```

and then pass $file into the Excel::import. 

Looking through here, there are a few other things I can see, so I'll add a few other comments

File: app/Exports/XlsformTemplateLanguageExport.php [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+use Maatwebsite\Excel\Concerns\WithBackgroundColor;
+
+class XlsformTemplateLanguageExport implements FromArray, WithHeadings, WithTitle, WithColumnWidths, WithStyles, WithBackgroundColor
+{
+    private $xlsformTemplateId;
+
+    public function __construct($templateId)
+    {
+        $this->xlsformTemplateId = $templateId;
+    }
```
[the remark this answers is not in the document]
Comment (dave-mills) [reply app/Exports/XlsformTemplateLanguageExport.php:27]: This would suggest a slightly different order for creating XlsformTemplateLanguage entries. Perhaps rather than creating a new entry when you upload a completed form translation, you instead create new entries at step 1. This implies:

- XlsformTemplateLanguage entries also belong to a team (the team that does the translation)
- we need a way of showing if a translation has been uploaded for the entry or not (so the team can see which ones they have completed. 

In the example above, maybe when the team picks Spanish, they might decide to use the Spanish Mexico version with no changes. In that case, we still need a way of showing that the team is using that language, even though its not "theirs" (they didn't upload it).

File: app/Imports/XlsformTemplateLanguageImport.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+use App\Models\LanguageStringType;
+use Maatwebsite\Excel\Concerns\ToModel;
+use Maatwebsite\Excel\Concerns\WithHeadingRow;
+
+class XlsformTemplateLanguageImport implements ToModel, WithHeadingRow
+{
+    protected $templateId;
+    protected $templateLanguageId;
+
+    public function __construct($templateId, $templateLanguageId)
```
Comment (dave-mills) [inline app/Imports/XlsformTemplateLanguageImport.php:16]: Probably only need to pass in the XlsformTemplateLanguage entry here, as the template can be got from the relationship?

File: app/Imports/XlsformTemplateLanguageImport.php [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    {
+        $this->templateId = $templateId;
+        $this->templateLanguageId = $templateLanguageId;
+    }
+
+    public function model(array $row)
```
Comment (dave-mills) [inline app/Imports/XlsformTemplateLanguageImport.php:22]: Normally, the model() method asks you to return a `new Model()` class. As you're conditionally creating the new model here based on whether the surrvey row and language string type exists, I would use `onRow()` instead: https://docs.laravel-excel.com/3.1/imports/model.html#handling-persistence-on-your-own

File: app/Models/LanguageString.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+use App\Models\Language;
+use App\Models\SurveyRow;
+use App\Models\LanguageStringType;
+use App\Models\XlsformTemplateLanguage;
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\Relations\BelongsTo;
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+
+class LanguageString extends Model
```
Comment (dave-mills) [inline app/Models/LanguageString.php:13]: Probably also should have a relationship to the XlsformTemplateLanaguage entry?

File: app/Imports/XlsformTemplateImport.php [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+use App\Models\SurveyRow;
+use App\Models\LanguageString;
+use App\Models\LanguageStringType;
+use Illuminate\Support\Collection;
+use Illuminate\Support\Facades\Log;
+use App\Models\XlsformTemplateLanguage;
+use Maatwebsite\Excel\Concerns\ToCollection;
+use Maatwebsite\Excel\Concerns\WithHeadingRow;
+
+class XlsformTemplateImport implements ToCollection, WithHeadingRow
```
Comment (dave-mills) [inline app/Imports/XlsformTemplateImport.php:15]: We're going to need to split up the rows from the surveys worksheet and choices worksheet as they'll need to be handled differently when we come to rebuild the team's forms. So eventually we'll need this to be setup as a multiple worksheets import - but we can use basically the same approach for identifying translatable parts for both the survey and choice worksheets.

**Review by dave-mills — COMMENTED** [review]
Looks like a good setup. I really like the approach used to identify and extract the 'translatable' parts of the survey. Ive made a bunch of comments about specific parts - some refactoring suggestions, but also some thoughts about the overall workflow the teams will follow and the ordering of operations. 

Perhaps things you were already thinking about addressing once you got the core feature working, but thought it worth adding my thoughts while I had them.

File: app/Filament/Admin/Resources/XlsformTemplateResource/RelationManagers/XlsformTemplateLanguageRelationManager.php [hunk 6]
```diff
@@ -102,15 +103,51 @@ public function table(Table $table): Table
...
+                    
+                        // If there are missing translations, display an error and prevent the import
+                        if ($invalidRows->isNotEmpty()) {
+                            Notification::make()
+                                ->title('Upload unsuccessful')
+                                ->body('The translations file cannot be uploaded as some translations are missing')
+                                ->danger()
+                                ->send();
+                            return;
+                        }
```
Comment (dave-mills) [inline app/Filament/Admin/Resources/XlsformTemplateResource/RelationManagers/XlsformTemplateLanguageRelationManager.php:133]: Logic looks sound here, and it works for me locally. 

The one thing I'd suggest is turning this into custom rule function on the UploadField - that way we can show an inline message when the validation fails, and it won't close the modal (because it won't get to the action() function). See my commit

**Comment by dave-mills** [comment]
I think that last update works. The step-by-step approach is solid. I think there's a  few things to think about on how to present the workflow to users, but that's not for this PR. For me, this works and is a good approach 👍.

**Review by dave-mills — COMMENTED** [review]
What you've done with making the exports and imports TemplateLanguage specific looks good and works well. 

It looks like the main thing that's left is to fix the thing where a whole new set of SurveyRow entries are created when you change the template xlsform file, which means: 
- use updateOrCreate  for SurveyRow and LanguageString;
- check any existing language strings to see if they've changed and if they have, set strings from the languages not present in the file to empty;
- delete any SurveyRow and LanguageString entries that no longer exist in the form. 

Is that right?

**Comment by ciara-mc** [comment]
> What you've done with making the exports and imports TemplateLanguage specific looks good and works well.
> 
> It looks like the main thing that's left is to fix the thing where a whole new set of SurveyRow entries are created when you change the template xlsform file, which means:
> 
> * use updateOrCreate  for SurveyRow and LanguageString;
> * check any existing language strings to see if they've changed and if they have, set strings from the languages not present in the file to empty;
> * delete any SurveyRow and LanguageString entries that no longer exist in the form.
> 
> Is that right?

Yes I'm working on this, here is the workflow:

1. Check new survey rows against existing survey rows
	- if one is removed: delete survey row, delete all associated language strings
	- if one is added: create the survey row, create any language strings, add needs_update flag on all other template languages if there are new language strings

2. Check new language strings against existing language strings
	- if one is removed: delete the language string, find all other language strings of the same type for the survey row and delete
	- if one is added: create the language string, add needs_update flag on all other template languages
	- if one is changed: update the language string, find all other language strings of the same type for the survey row and set to empty, add needs_update flag on all other template languages

**Review by dave-mills — APPROVED** [review]
The `needs_update` function wasn't working, but I pushed a fix for it and now I think the logic is ok. The rest works well - this is good to merge in 👍
